domain/machine/state/state.go

```diff
@@ -909,33 +909,48 @@ func (st *State) SetLXDProfiles(ctx context.Context, mUUID string, profileNames
 	}
 
 	checkMachineExistsStmt, err := st.Prepare(`
-SELECT true AS &machineExists.exists
+SELECT uuid AS &machineUUID.uuid
 FROM   machine
-WHERE  machine.uuid = $machineUUID.uuid`, machineUUID{}, machineExists{})
+WHERE  machine.uuid = $machineUUID.uuid`, machineUUID{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	checkProfileExistsStmt, err := st.Prepare(`
+SELECT machine_uuid AS &lxdProfile.machine_uuid
+FROM   machine_lxd_profile
+WHERE  machine_uuid = $lxdProfile.machine_uuid
+AND    name = $lxdProfile.name`, lxdProfile{})
 	if err != nil {
 		return errors.Trace(err)
 	}
 
 	setLXDProfileStmt, err := st.Prepare(`
-INSERT INTO lxd_profile (*)
-VALUES      ($lxdProfile.*)
-ON CONFLICT DO NOTHING`, lxdProfile{})
+INSERT INTO machine_lxd_profile (*)
+VALUES      ($lxdProfile.*)`, lxdProfile{})
 	if err != nil {
 		return errors.Trace(err)
 	}
 
 	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		var exists machineExists
-		err = tx.Query(ctx, checkMachineExistsStmt, machineUUID{UUID: mUUID}).Get(&exists)
+		var machineExists machineUUID
+		err = tx.Query(ctx, checkMachineExistsStmt, machineUUID{UUID: mUUID}).Get(&machineExists)
 		if errors.Is(err, sqlair.ErrNoRows) {
 			return machineerrors.MachineNotFound
 		} else if err != nil {
 			return internalerrors.Errorf("checking if machine %q exists: %w", mUUID, err)
 		}
 
-		profilesToInsert := make([]lxdProfile, len(profileNames))
-		for i, profileName := range profileNames {
-			profilesToInsert[i] = lxdProfile{Name: profileName, MachineUUID: mUUID}
+		profilesToInsert := make([]lxdProfile, 0, len(profileNames))
+		for _, profileName := range profileNames {
```

Comment:
**suggestion:** You could have all the existing profiles at once, with a `WHERE name IN (...)` clauses.

This transaction would becames:

* Check machine exists
* Select existing profiles (a list)
* Exclude existing profiles from inserted one (pure go)
* Insert remaining profiles

Reply:
This has changed since your comment, we don't check for any previous lxd profiles and overwrite the whole list instead. We do this because we need to maintain the order and keep the same logic as in the legacy state.